[WIP] Asynchronous session service

Nateowami opened the pull request:
Opening this PR so it's easier to see what work is being done on my branch. I've created an async session service (a modified version of the original), migrate files one at a time to use it, ensure tests still pass, fix them if they don't, commit, repeat.

I started out trying to move everything over in one step, and that was a bad idea. This way is kind of slow, but it isolates issues, which has been extremely helpful in tracking down test failures.

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sillsdev/web-languageforge/128)
<!-- Reviewable:end -->

megahirt commented:
Reviewed 35 of 67 files at r1.
Review status: 35 of 67 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/angular-app/bellows/js/bellows.services.api.js, line 4 at r1](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128#-Kmozh114Mjct-7CC7TM:-Kmozh114Mjct-7CC7TN:b-in1jq8) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5ff11b8b16049a1d1f610313336a368ef13baac1/src/angular-app/bellows/js/bellows.services.api.js#L4)):*
> ```JavaScript
>   .service('apiService', ['jsonRpc', '$q', function (jsonRpc, $q) {
> 
>     // FIXME This same code exists in session service, but can't be used due to circular dependency.
> ```

you could expose a method `api.projectId()` which the session service could just call straight through to.  e.g. in the session:

`session.projectId = api.projectId;`

Then the logic would just be in one place.

---

*[src/angular-app/bellows/js/bellows.services.asyncsession.js, line 65 at r1](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128#-Kmow_rsABod0fb2ryDR:-Kmow_rsABod0fb2ryDS:bhma7sf) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5ff11b8b16049a1d1f610313336a368ef13baac1/src/angular-app/bellows/js/bellows.services.asyncsession.js#L65)):*
> ```JavaScript
> 
>     this.userId = fnFor('userId');
>     this.currentUserId = this.userId; // deprecated
> ```

great.  Let's move all instances of currentUserId over to userId.  Thanks.

---

*[src/angular-app/bellows/js/bellows.services.asyncsession.js, line 72 at r1](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128#-KmoxOsVkl5340yBJinN:-KmoxOsVkl5340yBJinO:b-lusbqa) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5ff11b8b16049a1d1f610313336a368ef13baac1/src/angular-app/bellows/js/bellows.services.asyncsession.js#L72)):*
> ```JavaScript
>     this.username = fnFor('username');
> 
>     // TODO Is it really necessary to have these on the instance as well?
> ```

I don't see a good reason to have them in both places.

---

*[src/angular-app/bellows/js/bellows.services.project.js, line 38 at r1](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128#-Kmoz7bToctU64T6xSyd:-Kmoz7bUaWepCwkX8EHo:bn53lyd) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5ff11b8b16049a1d1f610313336a368ef13baac1/src/angular-app/bellows/js/bellows.services.project.js#L38)):*
> ```JavaScript
>     this.sendJoinRequest = api.method('project_sendJoinRequest');
>     this.getJoinRequests = api.method('project_getJoinRequests');
>     this.getDto = api.method('project_management_dto'); // TODO this is just for names of reports right? If so rename
> ```

The project management dto was used by the projectmanagement app, but I think that got turned into a directive.  The idea is still to have a place where generic project management features exist, so I'm hesitant to rename it.

---

*[src/angular-app/languageforge/lexicon/lexicon.js, line 76 at r1](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128#-Kmp-OHJGitVmilCu7t4:-Kmp-OHJGitVmilCu7t5:b-silois) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5ff11b8b16049a1d1f610313336a368ef13baac1/src/angular-app/languageforge/lexicon/lexicon.js#L76)):*
> ```JavaScript
>     });
> 
>     $q.all([sessionService.getSession(), rightsService.getRights()]).then(function(data) {
> ```

$q.all()  Who knew?  Nice work.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128)*
<!-- Sent from Reviewable.io -->

megahirt commented:
Reviewed 23 of 67 files at r1.
Review status: 58 of 67 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/angular-app/bellows/apps/activity/activity.js, line 91 at r1](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128#-Kmp3VAemPPSi9_WOKNd:-Kmp3VAemPPSi9_WOKNe:bgngf9f) ([raw file](https://github.com/sillsdev/web-languageforge/blob/7d5d4434337e4cd44071a68bc4cc1268dc66e011/src/angular-app/bellows/apps/activity/activity.js#L91)):*
> ```JavaScript
>       $scope.filteredActivities = [];
>       angular.forEach($scope.activities, function (activity) {
>         if (activity.userRef && activity.userRef.id == sessionService.currentUserId() ||
> ```

session.userId() ?

---

*[src/angular-app/languageforge/lexicon/js/lexicon.services.rights.js, line 23 at r1](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128#-Kmp1QZArryq3XmrkQrD:-Kmp1QZArryq3XmrkQrE:boe63tq) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5ff11b8b16049a1d1f610313336a368ef13baac1/src/angular-app/languageforge/lexicon/js/lexicon.services.rights.js#L23)):*
> ```JavaScript
>      * allowed to perform the given operation.
>      */
>     function condition(allowArchived, domain, operation, projectOwnerAllowed) {
> ```

this is elegant and makes it easier to read what permissions map to what actions.  You've made it easier to change the policy for the overrides for permissions, such as send/receive in progress or project is owner.  Nice work.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128)*
<!-- Sent from Reviewable.io -->

megahirt commented:
Hi Nathaniel,

This is an epic PR with loads of refactoring and good work.  It's very important for us to get this into master ASAP to avoid conflicts and having other devs start testing it.  As soon as you've finished moving all session methods over to async, then we'll merge the PR and start over with another one.  Thanks!

---

Reviewed 9 of 67 files at r1.
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[src/angular-app/bellows/apps/changepassword/changepassword.js, line 12 at r1](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128#-Kmp3_u1tBku7jVLLMFR:-Kmp3_u1tBku7jVLLMFS:b-3x4l7l) ([raw file](https://github.com/sillsdev/web-languageforge/blob/7d5d4434337e4cd44071a68bc4cc1268dc66e011/src/angular-app/bellows/apps/changepassword/changepassword.js#L12)):*
> ```JavaScript
>       $scope.updatePassword = function () {
>         if ($scope.vars.password == $scope.vars.confirm_password) {
>           userService.changePassword(sessionService.currentUserId(), $scope.vars.password,
> ```

session.userId() ?  Probably a quick find in files for currentUserId() and replace with userId() will suffice.

---

*[src/Api/Model/Shared/Command/SessionCommands.php, line 35 at r1](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128#-Kmp491tGREKlO8vBLTA:-Kmp491tGREKlO8vBLTB:b-c35ty) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5ff11b8b16049a1d1f610313336a368ef13baac1/src/Api/Model/Shared/Command/SessionCommands.php#L35)):*
> ```PHP
>         if ($projectId) {
>             $project = ProjectModel::getById($projectId);
>             if (array_key_exists($userId, $project->users)) {
> ```

Is this here to fix a bug?  The reason I ask is that someday in the future we will implement public/anonymous access to projects (that enable it).  I'm undecided as to how to actually implement it, but this check could cause issues with the implementation.

---

*[src/Site/views/shared/container/xforge.navbarApp.js, line 14 at r1](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128#-Kmp4lRJI8Wov2wl7Nx0:-Kmp4lRKWDJxYjxbI2ll:b-5ga63b) ([raw file](https://github.com/sillsdev/web-languageforge/blob/7d5d4434337e4cd44071a68bc4cc1268dc66e011/src/Site/views/shared/container/xforge.navbarApp.js#L14)):*
> ```JavaScript
>   */
>   .controller('navbarController', ['$scope', 'modalService', '$location', 'sessionService',
>   function ($scope, modalService, $location, ss) {
> ```

Don't worry much about xforge.navbarApp.js  At this point I'm planning on removing it entirely...never shipped and not that useful.

---

*[test/app/bellows/public/signup/e2e/signup.spec.js, line 126 at r1](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128#-Kmp4zu59lyo6Hyo6xAH:-Kmp4zu6XrflEljYcvXt:b-ygcdwv) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5ff11b8b16049a1d1f610313336a368ef13baac1/test/app/bellows/public/signup/e2e/signup.spec.js#L126)):*
> ```JavaScript
>     loginPage.loginAsUser();
>     page.get();
>     browser.wait(expectedCondition.visibilityOf(projectsPage.createBtn), CONDITION_TIMEOUT); // flaky wait (can cause "document unloaded while waiting for result")
> ```

A document unloaded while waiting for result is purely a test framework error, not an actual error.  Still, it would be good to figure out how to never encounter them.  My opinion is that ignoring console errors that match "document unloaded while waiting for result" is the lesser evil as opposed to dealing with flaky tests that are performance dependent (and therefore require a browser wait condition)

---

*[test/app/bellows/userProfile/e2e/userprofile.spec.js, line 185 at r1](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128#-Kmp5XOjLtBGO5G2Jn4O:-Kmp5XOjLtBGO5G2Jn4P:burc17k) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5ff11b8b16049a1d1f610313336a368ef13baac1/test/app/bellows/userProfile/e2e/userprofile.spec.js#L185)):*
> ```JavaScript
>         // Login with new username and revert to original username
>         loginPage.get();
>         expect(loginPage.infoMessages.count()).toBe(1); // flaky assertion?
> ```

Could be.  Can we just do without the assertion entirely?

---

*[test/app/languageforge/lexicon/editor/e2e/editor-entry.spec.js, line 20 at r1](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128#-Kmp5fxiRIVRVXxK3Oas:-Kmp5fxiRIVRVXxK3Oat:bbf9y2f) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5ff11b8b16049a1d1f610313336a368ef13baac1/test/app/languageforge/lexicon/editor/e2e/editor-entry.spec.js#L20)):*
> ```JavaScript
> 
>   it('browse page has correct word count', function () {
>     expect(editorPage.browse.entriesList.count()).toEqual(editorPage.browse.getEntryCount()); // flaky assertion
> ```

how is this flaky?

---

*[test/app/languageforge/lexicon/newLexProject/e2e/lexicon-new-project.spec.js, line 131 at r1](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128#-Kmp5nOmbRfvyr5m4xJu:-Kmp5nOmbRfvyr5m4xJv:bbf9y2f) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5ff11b8b16049a1d1f610313336a368ef13baac1/test/app/languageforge/lexicon/newLexProject/e2e/lexicon-new-project.spec.js#L131)):*
> ```JavaScript
>         CONDITION_TIMEOUT);
>       expect(page.srCredentialsPage.loginOk.isDisplayed()).toBe(false);
>       expect(page.srCredentialsPage.credentialsInvalid.isDisplayed()).toBe(true); // flaky assertion
> ```

how is this flaky?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128#-:-Kmp5tmPJqOggrvSCihX:b-4whqi2)*
<!-- Sent from Reviewable.io -->

irahopkinson commented:
*[test/app/bellows/public/signup/e2e/signup.spec.js, line 126 at r1](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128#-Kmp4zu59lyo6Hyo6xAH:-Kmq4W81pKBWAPJnLHZ5:b-glp6vk) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5ff11b8b16049a1d1f610313336a368ef13baac1/test/app/bellows/public/signup/e2e/signup.spec.js#L126)):*
<details><summary><i>Previously, megahirt (Chris Hirt) wrote…</i></summary><blockquote>

A document unloaded while waiting for result is purely a test framework error, not an actual error.  Still, it would be good to figure out how to never encounter them.  My opinion is that ignoring console errors that match "document unloaded while waiting for result" is the lesser evil as opposed to dealing with flaky tests that are performance dependent (and therefore require a browser wait condition)
</blockquote></details>

You can see how I fixed this for the prior test on master https://github.com/sillsdev/web-languageforge/blob/master/test/app/bellows/public/signup/e2e/signup.spec.js#L115

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128)*
<!-- Sent from Reviewable.io -->

cambell-prince commented on src/angular-app/bellows/apps/userprofile/userprofile.js:
```diff
@@ -14,9 +14,9 @@ angular.module('userprofile', ['ui.bootstrap', 'bellows.services', 'palaso.ui.no
+  .controller('userProfileCtrl', ['$scope', 'userService', 'utilService',
   'silNoticeService', 'modalService', '$window',
-function ($scope, userService, ss, util, notice, modalService, $window) {
+function ($scope, userService, util, notice, modalService, $window) {
```
Good catch.

cambell-prince commented on src/angular-app/languageforge/lexicon/directive/dc-semanticdomain.js:
```diff
@@ -12,9 +12,8 @@ angular.module('palaso.ui.dc.semanticdomain', [])
       selectField: '&'
     },
-    controller: ['$scope', '$state', 'lexRightsService', function ($scope, $state, rights) {
+    controller: ['$scope', '$state', 'lexRightsService', function ($scope, $state, rightsService) {
```
Is this confined to being a LexiconRightsService?  If not then there's an opportunity to use it elsewhere, there's rights use above that uses the session service for rights.  No problem with this code, just pointing out the possible opportunity for consolidating.

Nateowami replied:
The `lexRightsService` uses the `sessionService`, adding additional logic, such as whether a project is archived, and whether a send and receive is in progress (thus disallowing certain functions that would otherwise be allowed).

cambell-prince commented on src/angular-app/languageforge/lexicon/js/lexicon.services.rights.js:
```diff
@@ -1,67 +1,59 @@
+      };
+    }
+
+    var Rights = (new function () {
```
The Rights helper would do well as part of a UserProject service.

Nateowami replied:
Indeed. I think the session service has grown into "stuff that's convenient to have globally" rather than strictly session data. Now that it does all this it could use some refactoring into multiple services.

Nateowami commented:
Review status: all files reviewed at latest revision, 19 unresolved discussions.

---

*[src/angular-app/bellows/apps/changepassword/changepassword.js, line 12 at r1](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128#-Kmp3_u1tBku7jVLLMFR:-KmybkEv7vVhx6BNp-fL:b-6oc78r) ([raw file](https://github.com/sillsdev/web-languageforge/blob/7d5d4434337e4cd44071a68bc4cc1268dc66e011/src/angular-app/bellows/apps/changepassword/changepassword.js#L12)):*
<details><summary><i>Previously, megahirt (Chris Hirt) wrote…</i></summary><blockquote>

session.userId() ?  Probably a quick find in files for currentUserId() and replace with userId() will suffice.
</blockquote></details>

Good catch. I need to do a global search for these. Added to my TODO list.

---

*[src/angular-app/bellows/js/bellows.services.api.js, line 4 at r1](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128#-Kmozh114Mjct-7CC7TM:-Kmyf8ytHgCqsaFPNtM8:bp2xgi2) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5ff11b8b16049a1d1f610313336a368ef13baac1/src/angular-app/bellows/js/bellows.services.api.js#L4)):*
<details><summary><i>Previously, megahirt (Chris Hirt) wrote…</i></summary><blockquote>

you could expose a method `api.projectId()` which the session service could just call straight through to.  e.g. in the session:

`session.projectId = api.projectId;`

Then the logic would just be in one place.
</blockquote></details>

That's a good idea. I think I'll do that.

---

*[src/angular-app/bellows/js/bellows.services.asyncsession.js, line 72 at r1](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128#-KmoxOsVkl5340yBJinN:-KmyhWMJCCB51HZeyR9p:b-b54eec) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5ff11b8b16049a1d1f610313336a368ef13baac1/src/angular-app/bellows/js/bellows.services.asyncsession.js#L72)):*
<details><summary><i>Previously, megahirt (Chris Hirt) wrote…</i></summary><blockquote>

I don't see a good reason to have them in both places.
</blockquote></details>

I'll move them all to the static session (rather than the singleton instance).

---

*[src/angular-app/bellows/js/bellows.services.project.js, line 38 at r1](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128#-Kmoz7bToctU64T6xSyd:-Kmyhl7k2aqfexFtmlyJ:bmudmh9) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5ff11b8b16049a1d1f610313336a368ef13baac1/src/angular-app/bellows/js/bellows.services.project.js#L38)):*
<details><summary><i>Previously, megahirt (Chris Hirt) wrote…</i></summary><blockquote>

The project management dto was used by the projectmanagement app, but I think that got turned into a directive.  The idea is still to have a place where generic project management features exist, so I'm hesitant to rename it.
</blockquote></details>

Looking back I'm not completely sure why I found this name so confusing. I spent a good while trying to figure out which api method is used for fetching project data (with the idea that project data currently in the session needs to be moved there). I didn't end up finding it.

The name can stay; it describes what the procedure is for, which makes sense with RPC, but wouldn't so much with a REST api (which would describe what it contains). I'll remove the comment in a future commit.

---

*[src/Api/Model/Shared/Command/SessionCommands.php, line 35 at r1](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128#-Kmp491tGREKlO8vBLTA:-Kmz0zA2BJTYvxlf8usk:bacl8xf) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5ff11b8b16049a1d1f610313336a368ef13baac1/src/Api/Model/Shared/Command/SessionCommands.php#L35)):*
<details><summary><i>Previously, megahirt (Chris Hirt) wrote…</i></summary><blockquote>

Is this here to fix a bug?  The reason I ask is that someday in the future we will implement public/anonymous access to projects (that enable it).  I'm undecided as to how to actually implement it, but this check could cause issues with the implementation.
</blockquote></details>

I added this check for an edge case. In order to provide the user's permissions in the DTO, it has to look up the user's role on the project. If a user is removed from a project (specifically, the most recent project the user has visited) this causes an invalid index (can't remember exact name) exception. Previously this couldn't be triggered through the UI, because when a user visits a non-project page, no project data is sent. What what we should really be doing is getting the projectId from the client (rather than the server's session). I'm not clear whether we're ready to do that yet. Since most (all?) api-using services now use the apiService, most api calls now include the projectId.

---

*[test/app/bellows/public/signup/e2e/signup.spec.js, line 126 at r1](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128#-Kmp4zu59lyo6Hyo6xAH:-Kmz3xEklcIEGOyQch4r:b8kewvc) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5ff11b8b16049a1d1f610313336a368ef13baac1/test/app/bellows/public/signup/e2e/signup.spec.js#L126)):*
<details><summary><i>Previously, irahopkinson (Ira Hopkinson) wrote…</i></summary><blockquote>

You can see how I fixed this for the prior test on master https://github.com/sillsdev/web-languageforge/blob/master/test/app/bellows/public/signup/e2e/signup.spec.js#L115
</blockquote></details>

The condition check is executed in the browser, and the browser unloaded while the function was running, so WebDriver blows up. It's not a console error, so it can't be ignored as one. Wait conditions are effective within a single page, but most of them have to execute checks in the browser, so they break if a page unloads.
I just tried adding a urlContains check as Ira suggested and it works wonderfully. Looking at the implementation, it appears it doesn't have to run a check inside the browser, so WebDriver can handle it fine. Most of the flaky test issues I've had recently deal with waiting for a client-side redirect, so this should solve quite a few of them.
I've been adding comments to flaky tests so that a) we know where we need to fix them, and b) when a test fails I can quickly check whether it's known to be flaky and can be ignored (this has saved me from *tons* of test reruns).

---

*[test/app/bellows/userProfile/e2e/userprofile.spec.js, line 185 at r1](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128#-Kmp5XOjLtBGO5G2Jn4O:-Kmz9RkQbq2JWxdzKBVT:b-dh5hzs) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5ff11b8b16049a1d1f610313336a368ef13baac1/test/app/bellows/userProfile/e2e/userprofile.spec.js#L185)):*
<details><summary><i>Previously, megahirt (Chris Hirt) wrote…</i></summary><blockquote>

Could be.  Can we just do without the assertion entirely?
</blockquote></details>

I suppose we could. Trying it manually, this always works, but when I pause the tests here, when the tests fail, there really is no message shown. I wonder whether it is a server-side race condition between two requests.

---

*[test/app/languageforge/lexicon/editor/e2e/editor-entry.spec.js, line 20 at r1](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128#-Kmp5fxiRIVRVXxK3Oas:-Kmz9nSZ17iDiD0BTczI:b-liaer0) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5ff11b8b16049a1d1f610313336a368ef13baac1/test/app/languageforge/lexicon/editor/e2e/editor-entry.spec.js#L20)):*
<details><summary><i>Previously, megahirt (Chris Hirt) wrote…</i></summary><blockquote>

how is this flaky?
</blockquote></details>

It looks like getEntryCount() uses an expectedCondition, while the other does not. So this must be a race condition.

---

*[test/app/languageforge/lexicon/newLexProject/e2e/lexicon-new-project.spec.js, line 131 at r1](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128#-Kmp5nOmbRfvyr5m4xJu:-KmzAmjuo1za_TEjnVa6:b-k9w9lz) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5ff11b8b16049a1d1f610313336a368ef13baac1/test/app/languageforge/lexicon/newLexProject/e2e/lexicon-new-project.spec.js#L131)):*
<details><summary><i>Previously, megahirt (Chris Hirt) wrote…</i></summary><blockquote>

how is this flaky?
</blockquote></details>

I don't know. Seems like the expectedCondition should ensure this always passes if it reaches that far. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128)*
<!-- Sent from Reviewable.io -->

Nateowami commented on src/Site/views/shared/container/xforge.navbarApp.js:
```diff
@@ -10,49 +10,34 @@ angular.module('xforge.navbarApp', ['bellows.services', 'ui.bootstrap', 'pascalp
   */
-  .controller('navbarController', ['$scope', 'modalService', '$location', 'sessionService',
-  function ($scope, modalService, $location, ss) {
+  .controller('navbarController', ['$scope', 'modalService', '$location', 'asyncSession', '$window',
```
Having two session services right now is purely for the sake of ease of porting dependencies to use an asynchronous session service. My plan has been to delete the sessionService and rename asyncSession to sessionService with a find and replace. I'm not planning to use asyncSession as a permanent name, just during work in progress while the sessionService name is already taken.

Nateowami commented on src/Site/views/shared/container/xforge.navbarApp.js:
```diff
@@ -10,49 +10,34 @@ angular.module('xforge.navbarApp', ['bellows.services', 'ui.bootstrap', 'pascalp
-        helpFilePathsAvailable = ss.session.helps.filePaths;
-      }
-
+      var appName = $window.location.pathname.split('/')[2];
```
Searching globally for `location.pathname.split` this is the only instance I find (it's possible it's implemented differently elsewhere). Chris intends to delete the navbarApp, but otherwise it would make sense to move this to some service that provides metadata.

megahirt commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 53 of 55 files at r2.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[src/angular-app/bellows/js/bellows.services.project.js, line 38 at r1](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128#-Kmoz7bToctU64T6xSyd:-Kn8uIC6VOWRcoU_jTL2:biwj5kt) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5ff11b8b16049a1d1f610313336a368ef13baac1/src/angular-app/bellows/js/bellows.services.project.js#L38)):*
<details><summary><i>Previously, Nateowami (Nathaniel Paulus) wrote…</i></summary><blockquote>

Looking back I'm not completely sure why I found this name so confusing. I spent a good while trying to figure out which api method is used for fetching project data (with the idea that project data currently in the session needs to be moved there). I didn't end up finding it.

The name can stay; it describes what the procedure is for, which makes sense with RPC, but wouldn't so much with a REST api (which would describe what it contains). I'll remove the comment in a future commit.
</blockquote></details>

I think what we're missing is exactly that.  A generic DTO that returns the entire project object, and specific apps (lexicon, sfchecks) can inherit from that and include additional app specific information.

---

*[src/Site/views/shared/container/xforge.navbarApp.js, line 20 at r1](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128#-KmBbeFC9gAe7btfIiV9-r1-20:-Kn8wo_ZpAp3lWMBoIEi:bl7bkht) ([raw file](https://github.com/sillsdev/web-languageforge/blob/5ff11b8b16049a1d1f610313336a368ef13baac1/src/Site/views/shared/container/xforge.navbarApp.js#L20)):*
<details><summary><i>Previously, Nateowami (Nathaniel Paulus) wrote…</i></summary><blockquote>

Searching globally for `location.pathname.split` this is the only instance I find (it's possible it's implemented differently elsewhere). Chris intends to delete the navbarApp, but otherwise it would make sense to move this to some service that provides metadata. 
</blockquote></details>

It turns out that the client side doesn't make use of the appName at all as far as I remember, which is why I remember having to get it out of the URL.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/web-languageforge/128#-:-Kn8xazxjxErnK74XnFb:bnfp4nl)*
<!-- Sent from Reviewable.io -->
